packages/utils/telemetry-utils/src/sampledTelemetryHelper.ts

```diff
@@ -0,0 +1,143 @@
+
+    /**
+     * @param eventBase -
+     * Custom properties to include in the telemetry performance event when it is written.
+     * @param logger -
+     * The logger to use to write the telemetry performance event.
+     * @param sampleThreshold -
+     * Telemetry performance events will be generated every time we hit this many executions of the code block.
+     * @param includeAggregateMetrics -
+     * If set to `true`, the telemetry performance event will include aggregated metrics (total duration, min duration,
+     * max duration) for all the executions in between generated events.
+     * @param perBucketProperties -
+     * Map of strings that represent different buckets (which can be specified when calling the 'measure' method), to
+     * properties which should be added to the telemetry event for that bucket. If a bucket being measured does not
+     * have an entry in this map, no additional properties will be added to its telemetry events.
+     */
+    public constructor(
+        private readonly eventBase: ITelemetryGenericEvent,
+        private readonly logger: ITelemetryLogger,
+        private readonly sampleThreshold: number,
+        private readonly includeAggregateMetrics: boolean = false,
+        private readonly perBucketProperties = new Map<string, ITelemetryProperties>()) {
+    }
+
+    /**
+     * @param codeToMeasure -
+     * The code to be executed and measured.
+     * @param bucket -
+     * A key to track executions of the code block separately. Each different value of this parameter has a separate
+     * set of executions and metrics tracked by the class. If no such distinction needs to be made, do not provide a
+     * value.
+     * @returns Whatever the passed-in code block returns.
+     */
+    public measure<T>(codeToMeasure: () => T, bucket: string = ""): T {
+        const start = performance.now();
+        const returnValue = codeToMeasure();
+        const duration = performance.now() - start;
+
+        let m = this.measurementsMap.get(bucket);
+        if (m === undefined) {
+            m = { count: 0, duration: -1 };
+            this.measurementsMap.set(bucket, m);
+        }
+        m.count++;
+        m.duration = duration;
+
+        if (this.includeAggregateMetrics) {
+            m.totalDuration = (m.totalDuration ?? 0) + duration;
+            m.minDuration = Math.min(m.minDuration ?? duration, duration);
+            m.maxDuration = Math.max(m.maxDuration ?? 0, duration);
+        }
+
+        if (m.count >= this.sampleThreshold) {
+            this.flushBucket(bucket);
+        }
+
+        return returnValue;
+    }
+
+    private flushBucket(bucket: string) {
+        const measurements = this.measurementsMap.get(bucket);
+        if (measurements === undefined) {
+            return;
+        }
+
+        if (measurements.count !== 0) {
+            const telemetryEvent: ITelemetryPerformanceEvent = {
+                ...this.eventBase,
+                ...measurements,
```

Comment:
```suggestion
                ...bucketProperties,
                ...measurements,
```
if you get before this, you can just expand the elements in. i'm pretty sure we can spread undefined with no problem